web-ui/src/app/components/tree-visualizer/tree-visualizer.js

```diff
@@ -446,41 +447,60 @@ var jQuery = require('jquery');
                 var targetLink = tree.find("a.hovered");
                 if (targetLink.length) {
                     var tooltip = tree.next(".tv-tooltip"),
+                        arrow = tooltip.siblings(".arrow"),
                         targetRect = targetLink[0].getBoundingClientRect(),
                         treeRect = tree[0].getBoundingClientRect(),
+                        arrowRect = arrow[0].getBoundingClientRect(),
                         targetV = {
-                            right: $window.outerWidth() - targetRect.right,
-                            bottom: $window.outerHeight() - targetRect.bottom,
+                            right: $window.width() - targetRect.right,
+                            bottom: $window.innerHeight() - targetRect.bottom,
                         },
                         treeV = {
-                            right: $window.outerWidth() - treeRect.right,
-                            bottom: $window.outerHeight() - treeRect.bottom,
+                            right: $window.width() - treeRect.right,
+                            bottom: $window.innerHeight() - treeRect.bottom,
                         };
-
+                    // The maximum to the right the tooltip can be positioned without overlapping the viewport
+                    var rightestLeft = parseInt($window.width() - tooltip.outerWidth() - 20, 10)
                     var tooltipV = {
-                        left: parseInt(targetRect.left + (targetRect.width / 2) - (tooltip.outerWidth() / 2) + 8, 10) + "px",
+                        left: Math.max(10, Math.min(rightestLeft, parseInt(targetRect.left + (targetRect.width / 2) - (tooltip.outerWidth() / 2) + 8, 10))) + "px",
```

Comment:
The tooltip should be allowed to go more to the right if the node itself is moved to the right. Otherwise it looks really weird when scrolling the tree out of view:

![image](https://user-images.githubusercontent.com/24268578/45888167-b2246b80-bdbd-11e8-9b7a-e3240f69af7e.png)

Reply:
Hmm yeah this is slightly silly. Another solution would be to have the node (and thus the tooltip) fade out earlier. Then the tooltip will not cross the viewport (which in this case would happen) Which do you prefer?

Reply:
Good point, that might be even better!